### PR TITLE
fix the same handlerFunc

### PR DESCRIPTION
**Pull request by woniu317**
If one handlerFunc has been used for two or more routes, the transaction name is wrong.

```
func TestMiddlewareTransportName(t *testing.T) {
	debugOutput.Reset()
	tracer, transport := transporttest.NewRecorderTracer()
	defer tracer.Close()

	e := gin.New()
	e.Use(apmgin.Middleware(e, apmgin.WithTracer(tracer)))
	e.GET("/hello/:name", handleHello)
	e.GET("/world/:name", handleHello)

	w := httptest.NewRecorder()
	req, _ := http.NewRequest("GET", "http://server.testing/hello/isbel", nil)
	req.Header.Set("User-Agent", "apmgin_test")
	req.RemoteAddr = "client.testing:1234"
	e.ServeHTTP(w, req)
	tracer.Flush(nil)

	payloads := transport.Payloads()
	transaction := payloads.Transactions[0]
	assert.Equal(t, "GET /hello/:name", transaction.Name)
}

func handleHello(c *gin.Context) {
	c.String(http.StatusOK, "Hello, %s!", c.Param("name"))
}
```

The request is '/hello/:name', but the transaction.Name is '/world/:name'.

### CODE REVIEW

**Comment by axw** [comment]
@woniu317 thanks for the PR! Can you please add a test illustrating the problem?

Also, we'll need you to [sign the CLA](https://www.elastic.co/contributor-agreement) before we can accept any contributions.

**Comment by woniu317** [comment]
> @woniu317 thanks for the PR! Can you please add a test illustrating the problem?
> 
> Also, we'll need you to [sign the CLA](https://www.elastic.co/contributor-agreement) before we can accept any contributions.

```
func TestMiddlewareTransportName(t *testing.T) {
	debugOutput.Reset()
	tracer, transport := transporttest.NewRecorderTracer()
	defer tracer.Close()

	e := gin.New()
	e.Use(apmgin.Middleware(e, apmgin.WithTracer(tracer)))
	e.GET("/hello/:name", handleHello)
	e.GET("/world/:name", handleHello)

	w := httptest.NewRecorder()
	req, _ := http.NewRequest("GET", "http://server.testing/hello/isbel", nil)
	req.Header.Set("User-Agent", "apmgin_test")
	req.RemoteAddr = "client.testing:1234"
	e.ServeHTTP(w, req)
	tracer.Flush(nil)

	payloads := transport.Payloads()
	transaction := payloads.Transactions[0]
	assert.Equal(t, "GET /hello/:name", transaction.Name)
}

func handleHello(c *gin.Context) {
	c.String(http.StatusOK, "Hello, %s!", c.Param("name"))
}

```

The request is '/hello/:name', but the transaction.Name is '/world/:name'.

**Comment by simitt** [comment]
Thanks for the PR @woniu317. As my colleague mentioned, before we can accept the PR, can you please complete these two steps:
* sign the CLA
* add the testcase that you used for demonstrating the bug, to the code.

**Comment by woniu317** [comment]
> Thanks for the PR @woniu317. As my colleague mentioned, before we can accept the PR, can you please complete these two steps:
> 
> * sign the CLA
> * add the testcase that you used for demonstrating the bug, to the code.

The protocol has been signed. How to automatically associate with github?
image

**Comment by simitt** [comment]
@woniu317 the CLA needs to be signed with the same user/email-address that what was used for the [commits](https://github.com/elastic/apm-agent-go/pull/979/commits). Is that also your user, in which case, could you please use that user to sign?

File: module/apmgin/middleware.go [hunk 1]
```diff
@@ -77,27 +73,17 @@ func (m *middleware) handle(c *gin.Context) {
...
-		m.routeMap = rm
 	})
 
-	var requestName string
-	handlerName := c.HandlerName()
-	if routeInfo, ok := m.routeMap[c.Request.Method][handlerName]; ok {
-		requestName = routeInfo.transactionName
-	} else {
+	requestName := c.Request.Method + " " + c.FullPath()
+	if _, ok := m.routeMap[requestName]; !ok {
```
Comment (simitt) [inline module/apmgin/middleware.go:83]: Building on top of the proposed changes, the solution can be further simplified by removing the `routeMap` completely and only check that the `fullPath` is not empty; the map does not provide any adavantage any more when not using the handler. 
Something similar to 

```
	var requestName string
	if fullPath := c.FullPath(); fullPath != "" {
		requestName = c.Request.Method + " " + c.FullPath()
	} else {
		requestName = apmhttp.UnknownRouteRequestName(c.Request)
	}
```

**Review by simitt — CHANGES_REQUESTED** [review]

**Comment by axw** [comment]
@woniu317 in case it's not clear, you have commits associated with a different email address to the one that you used to sign the CLA:

image

If you can squash and recreate those commits with your `woniu317` email/user, then the CLA checker should pass.

**Comment by woniu317** [comment]
see https://github.com/elastic/apm-agent-go/pull/983

**Pull request closed without merging**